web/courses.go

```diff
@@ -227,11 +227,13 @@ func (s *AutograderService) getAllLabs(request *pb.LabRequest) ([]*pb.LabResultL
 
 	// populate cache map with student labs, filtering the latest submissions for every assignment
 	for _, lab := range allLabs {
-		_, ok := labCache[lab.GetUserID()]
+		// submission record has only one ID: user ID or group ID
+		labID := lab.GetUserID() + lab.GetGroupID()
```

Comment:
This seems to me to be a potential source of a bug. Don’t user ids and group ids have overlapping domains? At least reading the code it is not obvious to me that this would be safe, and not return result of student’s 3 cache, when accessing the cache as group 3.

Wouldn’t it be better to do a bit wise or with a high bit set, if the group id > 0. E.g. the highest bit in the 64-bit id space?

Reply:
Yeah, it might be even better to split the logic for group and individual labs.

Reply:
Done.